strawberry/experimental/pydantic/utils.py

```diff
@@ -36,3 +43,61 @@ def get_private_fields(cls: Type) -> List[dataclasses.Field]:
             private_fields.append(field)
 
     return private_fields
+
+
+@dataclasses.dataclass()
+class DataclassCreationFields:
+    """Fields required for the fields parameter of make_dataclass"""
+
+    name: str
+    type_annotation: Any
+    field: dataclasses.Field
+
+    @property
+    def to_tuple(self) -> Tuple[str, Any, dataclasses.Field]:
+        # fields parameter wants (name, type, Field)
+        return (self.name, self.type_annotation, self.field)
+
+
+def sort_creation_fields(
+    fields: List[DataclassCreationFields],
+) -> List[DataclassCreationFields]:
+    """
+    Sort fields so that fields with missing defaults go first
+    because dataclasses require that fields with no defaults are defined
+    first
+    """
+    missing_default: List[DataclassCreationFields] = []
+    has_default: List[DataclassCreationFields] = []
+    for model_field in fields:
+        if (
+            model_field.field.default is dataclasses.MISSING
+            and model_field.field.default_factory is dataclasses.MISSING  # type: ignore
+        ):
+            missing_default.append(model_field)
+        else:
+            has_default.append(model_field)
+    return missing_default + has_default
```

Comment:
nice! thanks for moving these into a different file!

Reply:
Slightly shorter version
```python
    return sorted(
        fields,
        key=lambda model_field: (
            model_field.field.default is dataclasses.MISSING
            and model_field.field.default_factory is dataclasses.MISSING
        ),
        reverse=True
    )
```

Or with a function to keep annotations.
```python
    def is_missing_default(model_field: DataclassCreationFields):
        """Check if field has defaults."""
        return (
            model_field.field.default is dataclasses.MISSING
            and model_field.field.default_factory is dataclasses.MISSING
        )
    return sorted(fields, key=is_missing_default, reverse=True)
```

Reply:
I like the second one, great suggestion 😊

Reply:
Another option:
```python
    def has_default(model_field: DataclassCreationFields):
        """Check if field has defaults."""
        return (
            model_field.field.default is not dataclasses.MISSING
            or model_field.field.default_factory is not dataclasses.MISSING
        )
    return sorted(fields, key=has_default)
```